README.md
# Bit
Gtk3 Python Editor for the Micro::Bit

loosley based on the work done on Mu (codewith.mu)


debian/changelog
bit (1.0) unstable; urgency=low

  * Initial packaging

 -- Alexander Brown  Fri, 08 Apr 2016 08:56:35 +0100


bit.py
#!/usr/bin/python3

import sys
import gi
gi.require_version('Gtk', '3.0')
from gi.repository import GLib, Gio, Gtk, GObject
from gi.repository import GtkSource

# App Menu
MENU_XML="""
<?xml version="1.0" encoding="UTF-8"?>
<interface>
  <menu id="app-menu">
    <section>
      <item>
        <attribute name="action">app.about</attribute>
        <attribute name="label" translatable="yes">_About</attribute>
      </item>
      <item>
        <attribute name="action">app.quit</attribute>
        <attribute name="label" translatable="yes">_Quit</attribute>
        <attribute name="accel">&lt;Primary&gt;q</attribute>
    </item>
    </section>
  </menu>
</interface>
"""

class TabLabel(Gtk.Box):
    __gsignals__ = {
        "close-clicked": (GObject.SIGNAL_RUN_FIRST, GObject.TYPE_NONE, ()),
    }
    def __init__(self, label_text):
        Gtk.Box.__init__(self)
        self.set_orientation(Gtk.Orientation.HORIZONTAL)
        self.set_spacing(5) # spacing: [icon|5px|label|5px|close]  
             
        # label
        label = Gtk.Label(label_text)
        self.pack_start(label, True, True, 0)
       
        # close button
        button = Gtk.Button()
        button.set_relief(Gtk.ReliefStyle.NONE)
        button.set_focus_on_click(False)
        button.add(Gtk.Image.new_from_stock(Gtk.STOCK_CLOSE, Gtk.IconSize.MENU))
        button.connect("clicked", self.button_clicked)
        data =  ".button {\n" \
                "-GtkButton-default-border : 0px;\n" \
                "-GtkButton-default-outside-border : 0px;\n" \
                "-GtkButton-inner-border: 0px;\n" \
                "-GtkWidget-focus-line-width : 0px;\n" \
                "-GtkWidget-focus-padding : 0px;\n" \
                "padding: 0px;\n" \
                "}"
        provider = Gtk.CssProvider()
        provider.load_from_data(bytes(data, "ascii"))
        # 600 = GTK_STYLE_PROVIDER_PRIORITY_APPLICATION
        button.get_style_context().add_provider(provider, 600)
        self.pack_start(button, False, False, 0)
       
        self.show_all()
   
    def button_clicked(self, button, data=None):
        self.emit("close-clicked")

class BitWin(Gtk.ApplicationWindow):

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.header = Gtk.HeaderBar()
        self.header.set_show_close_button(True)
        self.header.set_title("Untitled")
        self.header.set_subtitle("Bit")

        btn_open  = Gtk.Button('Open')
        btn_save  = Gtk.Button('Save')
        btn_new   = Gtk.Button.new_from_icon_name('tab-new-symbolic', Gtk.IconSize.SMALL_TOOLBAR)
        btn_flash = Gtk.Button.new_from_icon_name('media-playback-start-symbolic', Gtk.IconSize.SMALL_TOOLBAR)

        self.header.pack_start(btn_open)
        self.header.pack_start(btn_new)
        self.header.pack_end(btn_save)
        self.header.pack_end(btn_flash)
        
        btn_new.connect("clicked", self.on_new_clicked, self)
        btn_open.connect("clicked", self.on_open_clicked, self)
        btn_save.connect("clicked", self.on_save_clicked, self)
        btn_flash.connect("clicked", self.on_flash_clicked, self)

        self.box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL)

        #toolbutton = Gtk.ToolButton(stock_id=Gtk.STOCK_NEW)
        #toolbutton.set_is_important(True)
        #toolbutton.connect("clicked", self.on_new_clicked, self)

        toolbar = Gtk.Toolbar()
        toolbar.get_style_context().add_class(Gtk.STYLE_CLASS_PRIMARY_TOOLBAR)
        #toolbar.insert(toolbutton, 0)

        self.box.add(toolbar)

        self.notebook = Gtk.Notebook()

        self.create_sourceview()
        self.box.pack_start(self.notebook, True, True, 0)
        
        self.add(self.box)
        self.set_default_size(800,600)
        self.set_titlebar(self.header)
        self.set_icon_name("applications-development")
        self.show_all()

    def create_sourceview(self, file = "/home/pi/projects/Untitled Folder/template.py"):

        filename = file.split('/')
        filename = filename[len(filename)-1]

        lm = GtkSource.LanguageManager.new()
        language = lm.guess_language(file, None)

        buffer = GtkSource.Buffer()

        if language:
            buffer.set_highlight_syntax(True)
            buffer.set_language(language)
            print(language)
        else:
            print('No language found for file "%s"' % file)
            buffer.set_highlight_syntax(False)

        source_file = GtkSource.File()
        source_file.set_location(Gio.File.new_for_path(file))
        source_file_loader = GtkSource.FileLoader().new(buffer, source_file);
        source_file_loader.load_async(GLib.PRIORITY_DEFAULT, None, None, None, None, None)

        # style list /usr/share/gtksourceview-3.0/
        sourceview = GtkSource.View.new_with_buffer(buffer)
        sourceview.set_auto_indent(True)
        sourceview.set_indent_on_tab(True)
        sourceview.set_show_line_numbers(True)
        sourceview.set_highlight_current_line(True)
        sourceview.set_smart_home_end(True)

        scrolledwindow = Gtk.ScrolledWindow()
        scrolledwindow.add(sourceview)

        tab_label = TabLabel(filename)
        tab_label.connect("close-clicked", self.on_close_clicked, self.notebook, scrolledwindow)
        
        self.notebook.append_page(scrolledwindow, tab_label)
        self.notebook.set_tab_reorderable(scrolledwindow, True)
        scrolledwindow.show_all()
        tab_label.show_all()
        self.notebook.set_current_page(self.notebook.page_num(scrolledwindow))

    def on_maximize_toggle(self, action, value):
        action.set_state(value)
        if value.get_boolean():
            self.maximize()
        else:
            self.unmaximize()

    def on_close_clicked(self, widget2, sender, widget):
        #get the page number of the tab we wanted to close
        pagenum = self.notebook.page_num(widget)
        #and close it
        self.notebook.remove_page(pagenum)

    def on_new_clicked(self, widget, data):
        self.create_sourceview()
        self.show_all()

    def on_save_clicked(self, widget, data):
        print("Save")
        
        
    def on_open_clicked(self, widget, data):
        print("Open")
        filechooserdialog = Gtk.FileChooserDialog("Please choose a file", self, Gtk.FileChooserAction.OPEN, (Gtk.STOCK_CANCEL, Gtk.ResponseType.CANCEL, Gtk.STOCK_OPEN, Gtk.ResponseType.OK))

        filter_py = Gtk.FileFilter()
        filter_py.set_name("Python files")
        filter_py.add_mime_type("text/x-python")
        filechooserdialog.add_filter(filter_py)

        filter_any = Gtk.FileFilter()
        filter_any.set_name("Any files")
        filter_any.add_pattern("*")
        filechooserdialog.add_filter(filter_any)
        
        response = filechooserdialog.run()
        if response == Gtk.ResponseType.OK:
            file = filechooserdialog.get_filename()
            self.create_sourceview(file)
        filechooserdialog.destroy()
        
    def on_flash_clicked(self, widget, data):
        print("Flash")
        

class BitApp(Gtk.Application):

    def __init__(self, *args, **kwargs):
        super().__init__(*args, application_id="org.zanderbrown.bit",
                         flags=Gio.ApplicationFlags.HANDLES_COMMAND_LINE,
                         **kwargs)
        self.window = None

        self.add_main_option("test", ord("t"), GLib.OptionFlags.NONE,
                             GLib.OptionArg.NONE, "Command line test", None)

    def do_startup(self):
        Gtk.Application.do_startup(self)

        action = Gio.SimpleAction.new("about", None)
        action.connect("activate", self.on_about)
        self.add_action(action)

        action = Gio.SimpleAction.new("quit", None)
        action.connect("activate", self.on_quit)
        self.add_action(action)

        builder = Gtk.Builder.new_from_string(MENU_XML, -1)
        self.set_app_menu(builder.get_object("app-menu"))

    def do_activate(self):
        # We only allow a single window and raise any existing ones
        if not self.window:
            # Windows are associated with the application
            # when the last one is closed the application shuts down
            self.window = BitWin(application=self, title="Bit")

        self.window.present()

    def do_command_line(self, command_line):
        options = command_line.get_options_dict()

        if options.contains("test"):
            # This is printed on the main instance
            print("Test argument recieved")

        self.activate()
        return 0

    def on_about(self, action, param):
        aboutdialog = Gtk.AboutDialog(transient_for=self.window, modal=True)
        aboutdialog.set_name("Bit")
        aboutdialog.set_version("1.0")
        aboutdialog.set_comments("Python Editor for Micro::Bit")
        aboutdialog.set_authors(["Alexander Brown"])
        aboutdialog.set_copyright("Copyright © 2016 Alexander Brown")
        aboutdialog.set_logo_icon_name("applications-development")
        aboutdialog.run()
        aboutdialog.destroy()

    def on_quit(self, action, param):
        self.quit()

if __name__ == "__main__":
    app = BitApp()
    app.run(sys.argv)


bit/logic.py
"""
Copyright 2016 Alexander Brown

Based on Mu:

Copyright (c) 2015-2016 Nicholas H.Tollervey and others (see the AUTHORS file).

Based upon work done for Puppy IDE by Dan Pope, Nicholas Tollervey and Damien
George.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""
import os
import os.path
import sys
import json
import logging
from PyQt5.QtWidgets import QMessageBox
from PyQt5.QtSerialPort import QSerialPortInfo
from mu.contrib import uflash, appdirs


#: USB product ID.
MICROBIT_PID = 516
#: USB vendor ID.
MICROBIT_VID = 3368
#: The user's home directory.
HOME_DIRECTORY = os.path.expanduser('~')
#: The default directory for Python scripts.
PYTHON_DIRECTORY = os.path.join(HOME_DIRECTORY, 'python')
#: The default directory for application data.
DATA_DIR = appdirs.user_data_dir('Bit', 'python')
#: The default directory for application logs.
LOG_DIR = appdirs.user_log_dir('Bit', 'python')
#: The path to the JSON file containing application settings.
SETTINGS_FILE = os.path.join(DATA_DIR, 'settings.json')
#: The path to the log file for the application.
LOG_FILE = os.path.join(LOG_DIR, 'Bit.log')


logger = logging.getLogger(__name__)


def find_microbit():
    """
    Returns the port for the first microbit it finds connected to the host
    computer. If no microbit is found, returns None.
    """
    available_ports = QSerialPortInfo.availablePorts()
    for port in available_ports:
        pid = port.productIdentifier()
        vid = port.vendorIdentifier()
        if pid == MICROBIT_PID and vid == MICROBIT_VID:
            port_name = port.portName()
            logger.info('Found micro:bit with portName: {}'.format(port_name))
            return port_name
    logger.warning('Could not find micro:bit.')
    logger.debug('Available ports:')
    logger.debug(['PID:{} VID:{} PORT:{}'.format(p.productIdentifier(),
                                                 p.vendorIdentifier(),
                                                 p.portName())
                 for p in available_ports])
    return None


class REPL:
    """
    Read, Evaluate, Print, Loop.

    Represents the REPL. Since the logic for the REPL is simply a USB/serial
    based widget this class only contains a reference to the associated port.
    """

    def __init__(self, port):
        if os.name == 'posix':
            # If we're on Linux or OSX reference the port is like this...
            self.port = "/dev/{}".format(port)
        elif os.name == 'nt':
            # On Windows simply return the port (e.g. COM0).
            self.port = port
        else:
            # No idea how to deal with other OS's so fail.
            raise NotImplementedError('OS not supported.')
        logger.info('Created new REPL object with port: {}'.format(self.port))


class Editor:
    """
    Application logic for the editor itself.
    """

    def __init__(self, view):
        logger.info('Setting up editor.')
        self._view = view
        self.repl = None
        self.theme = 'day'
        self.user_defined_microbit_path = None
        if not os.path.exists(PYTHON_DIRECTORY):
            logger.debug('Creating directory: {}'.format(PYTHON_DIRECTORY))
            os.makedirs(PYTHON_DIRECTORY)
        if not os.path.exists(DATA_DIR):
            logger.debug('Creating directory: {}'.format(DATA_DIR))
            os.makedirs(DATA_DIR)

    def restore_session(self):
        """
        Attempts to recreate the tab state from the last time the editor was
        run.
        """
        if os.path.exists(SETTINGS_FILE):
            logger.info('Restoring session from: {}'.format(SETTINGS_FILE))
            with open(SETTINGS_FILE) as f:
                old_session = json.load(f)
                logger.debug(old_session)
                if 'theme' in old_session:
                    self.theme = old_session['theme']
                if 'paths' in old_session:
                    for path in old_session['paths']:
                        try:
                            with open(path) as f:
                                text = f.read()
                        except FileNotFoundError:
                            pass
                        else:
                            self._view.add_tab(path, text)
        if not self._view.tab_count:
            py = 'from microbit import *\n\n# Write your code here :-)'
            self._view.add_tab(None, py)
        self._view.set_theme(self.theme)

    def flash(self):
        """
        Takes the currently active tab, compiles the Python script therein into
        a hex file and flashes it all onto the connected device.
        """
        logger.info('Flashing script')
        # Grab the Python script.
        tab = self._view.current_tab
        if tab is None:
            # There is no active text editor.
            return
        python_script = tab.text().encode('utf-8')
        logger.debug('Python script:')
        logger.debug(python_script)
        # Generate a hex file.
        python_hex = uflash.hexlify(python_script)
        logger.debug('Python hex:')
        logger.debug(python_hex)
        micropython_hex = uflash.embed_hex(uflash._RUNTIME, python_hex)
        # Determine the location of the BBC micro:bit. If it can't be found
        # fall back to asking the user to locate it.
        path_to_microbit = uflash.find_microbit()
        if path_to_microbit is None:
            # Has the path to the device already been specified?
            if self.user_defined_microbit_path:
                path_to_microbit = self.user_defined_microbit_path
            else:
                # Ask the user to locate the device.
                path_to_microbit = self._view.get_microbit_path(HOME_DIRECTORY)
                # Store the user's specification of the path for future use.
                self.user_defined_microbit_path = path_to_microbit
                logger.debug('User defined path to micro:bit: {}'.format(
                             self.user_defined_microbit_path))
        # Check the path and that it exists simply because the path maybe based
        # on stale data.
        logger.debug('Path to micro:bit: {}'.format(path_to_microbit))
        if path_to_microbit and os.path.exists(path_to_microbit):
            logger.debug('Flashing to device.')
            hex_file = os.path.join(path_to_microbit, 'micropython.hex')
            uflash.save_hex(micropython_hex, hex_file)
            message = 'Flashing "{}" onto the micro:bit.'.format(tab.label)
            information = ("When the yellow LED stops flashing the device"
                           " will restart and your script will run. If there"
                           " is an error, you'll see a helpful message scroll"
                           " across the device's display.")
            self._view.show_message(message, information, 'Information')
        else:
            # Reset user defined path since it's incorrect.
            self.user_defined_microbit_path = None
            # Try to be helpful... essentially there is nothing Mu can do but
            # prompt for patience while the device is mounted and/or do the
            # classic "have you tried switching it off and on again?" trick.
            # This one's for James at the Raspberry Pi Foundation. ;-)
            message = 'Could not find an attached BBC micro:bit.'
            information = ("Please ensure you leave enough time for the BBC"
                           " micro:bit to be attached and configured correctly"
                           " by your computer. This may take several seconds."
                           " Alternatively, try removing and re-attaching the"
                           " device or saving your work and restarting Mu if"
                           " the device remains unfound.")
            self._view.show_message(message, information)

    def add_repl(self):
        """
        Detect a connected BBC micro:bit and if found, connect to the
        MicroPython REPL and display it to the user.
        """
        logger.info('Starting REPL in UI.')
        if self.repl is not None:
            raise RuntimeError("REPL already running")
        mb_port = find_microbit()
        if mb_port:
            try:
                self.repl = REPL(port=mb_port)
                self._view.add_repl(self.repl)
                logger.info('REPL on port: {}'.format(mb_port))
            except IOError as ex:
                logger.error(ex)
                self.repl = None
                information = ("Click the device's reset button, wait a few"
                               " seconds and then try again.")
                self._view.show_message(str(ex), information)
            except Exception as ex:
                logger.error(ex)
        else:
            message = 'Could not find an attached BBC micro:bit.'
            information = ("Please make sure the device is plugged into this"
                           " computer.\n\nThe device must have MicroPython"
                           " flashed onto it before the REPL will work.\n\n"
                           "Finally, press the device's reset button and wait"
                           " a few seconds before trying again.")
            self._view.show_message(message, information)

    def remove_repl(self):
        """
        If there's an active REPL, disconnect and hide it.
        """
        if self.repl is None:
            raise RuntimeError("REPL not running")
        self._view.remove_repl()
        self.repl = None

    def toggle_repl(self):
        """
        If the REPL is active, close it; otherwise open the REPL.
        """
        if self.repl is None:
            self.add_repl()
        else:
            self.remove_repl()

    def toggle_theme(self):
        """
        Switches between themes (night or day).
        """
        if self.theme == 'day':
            self.theme = 'night'
        else:
            self.theme = 'day'
        logger.info('Toggle theme to: {}'.format(self.theme))
        self._view.set_theme(self.theme)

    def new(self):
        """
        Adds a new tab to the editor.
        """
        self._view.add_tab(None, '')

    def load(self):
        """
        Loads a Python file from the file system or extracts a Python sccript
        from a hex file.
        """
        path = self._view.get_load_path(PYTHON_DIRECTORY)
        logger.info('Loading script from: {}'.format(path))
        try:
            if path.endswith('.py'):
                # Open the file, read the textual content and set the name as
                # the path to the file.
                with open(path) as f:
                    text = f.read()
                name = path
            else:
                # Open the hex, extract the Python script therein and set the
                # name to None, thus forcing the user to work out what to name
                # the recovered script.
                with open(path) as f:
                    text = uflash.extract_script(f.read())
                name = None
        except FileNotFoundError:
            pass
        else:
            logger.debug(text)
            self._view.add_tab(name, text)

    def save(self):
        """
        Save the content of the currently active editor tab.
        """
        tab = self._view.current_tab
        if tab is None:
            # There is no active text editor so abort.
            return
        if tab.path is None:
            # Unsaved file.
            tab.path = self._view.get_save_path(PYTHON_DIRECTORY)
        if tab.path:
            # The user specified a path to a file.
            if not os.path.basename(tab.path).endswith('.py'):
                # No extension given, default to .py
                tab.path += '.py'
            with open(tab.path, 'w') as f:
                logger.info('Saving script to: {}'.format(tab.path))
                logger.debug(tab.text())
                f.write(tab.text())
            tab.setModified(False)
        else:
            # The user cancelled the filename selection.
            tab.path = None

    def zoom_in(self):
        """
        Make the editor's text bigger
        """
        self._view.zoom_in()

    def zoom_out(self):
        """
        Make the editor's text smaller.
        """
        self._view.zoom_out()

    def quit(self, *args, **kwargs):
        """
        Exit the application.
        """
        logger.info('Quitting')
        if self._view.modified:
            # Alert the user to handle unsaved work.
            msg = ('There is un-saved work, exiting the application will'
                   ' cause you to lose it.')
            result = self._view.show_confirmation(msg)
            if result == QMessageBox.Cancel:
                if args and hasattr(args[0], 'ignore'):
                    # The function is handling an event, so ignore it.
                    args[0].ignore()
                return
        paths = []
        for widget in self._view.widgets:
            if widget.path:
                paths.append(widget.path)
        session = {
            'theme': self.theme,
            'paths': paths
        }
        logger.debug(session)
        with open(SETTINGS_FILE, 'w') as out:
            logger.debug('Saving session to: {}'.format(SETTINGS_FILE))
            json.dump(session, out, indent=2)
        sys.exit(0)


bit/app.py
#!/usr/bin/python3

import sys, gi, logging, os
gi.require_version('Gtk', '3.0')
from gi.repository import GLib, Gio, Gtk, GObject
from gi.repository import GtkSource

#from mu.logic import Editor, LOG_FILE, LOG_DIR
from bit.logic import *

logger = logging.getLogger(__name__)

def setup_logging():
    """
    Configure logging.
    """
    if not os.path.exists(LOG_DIR):
        os.makedirs(LOG_DIR)
    log_format = '%(name)s(%(funcName)s) %(levelname)s: %(message)s'
    logging.basicConfig(filename=LOG_FILE, filemode='w', format=log_format,
                        level=logging.DEBUG)
    print('Logging to {}'.format(LOG_FILE))



# App Menu
MENU_XML="""
<?xml version="1.0" encoding="UTF-8"?>
<interface>
  <menu id="app-menu">
    <section>
      <item>
        <attribute name="action">app.about</attribute>
        <attribute name="label" translatable="yes">_About</attribute>
      </item>
      <item>
        <attribute name="action">app.quit</attribute>
        <attribute name="label" translatable="yes">_Quit</attribute>
        <attribute name="accel">&lt;Primary&gt;q</attribute>
    </item>
    </section>
  </menu>
</interface>
"""

class TabLabel(Gtk.Box):
    __gsignals__ = {
        "close-clicked": (GObject.SIGNAL_RUN_FIRST, GObject.TYPE_NONE, ()),
    }
    def __init__(self, label_text):
        logger.info("Make TabLabel")
        Gtk.Box.__init__(self)
        self.set_orientation(Gtk.Orientation.HORIZONTAL)
        self.set_spacing(5) # spacing: [icon|5px|label|5px|close]  

        # icon
        self.spinner = Gtk.Spinner()
        self.pack_start(self.spinner, False, False, 0)
        
        # label
        label = Gtk.Label(label_text)
        self.text = label_text
        self.pack_start(label, True, True, 0)
       
        # close button
        button = Gtk.Button()
        button.set_relief(Gtk.ReliefStyle.NONE)
        button.set_focus_on_click(False)
        button.add(Gtk.Image.new_from_stock(Gtk.STOCK_CLOSE, Gtk.IconSize.MENU))
        button.connect("clicked", self.button_clicked)
        data =  ".button {\n" \
                "-GtkButton-default-border : 0px;\n" \
                "-GtkButton-default-outside-border : 0px;\n" \
                "-GtkButton-inner-border: 0px;\n" \
                "-GtkWidget-focus-line-width : 0px;\n" \
                "-GtkWidget-focus-padding : 0px;\n" \
                "padding: 0px;\n" \
                "}"
        provider = Gtk.CssProvider()
        provider.load_from_data(bytes(data, "ascii"))
        # 600 = GTK_STYLE_PROVIDER_PRIORITY_APPLICATION
        button.get_style_context().add_provider(provider, 600)
        self.pack_start(button, False, False, 0)
       
        self.show_all()
        self.spinner.set_visible(False)
   
    def button_clicked(self, button, data=None):
        self.emit("close-clicked")

    def get_text(self):
        return self.text

    def start_working(self):
        self.spinner.start()
        self.spinner.set_visible(True)

    def stop_working(self):
        self.spinner.stop()
        self.spinner.set_visible(False)

class BitWin(Gtk.ApplicationWindow):

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.header = Gtk.HeaderBar()
        self.header.set_show_close_button(True)
        self.header.set_title("Untitled")
        self.header.set_subtitle("Bit")

        btn_open  = Gtk.Button('Open')
        btn_save  = Gtk.Button('Save')
        btn_new   = Gtk.Button.new_from_icon_name('tab-new-symbolic', Gtk.IconSize.SMALL_TOOLBAR)
        btn_flash = Gtk.Button.new_from_icon_name('media-playback-start-symbolic', Gtk.IconSize.SMALL_TOOLBAR)

        self.header.pack_start(btn_open)
        self.header.pack_start(btn_new)
        self.header.pack_end(btn_save)
        self.header.pack_end(btn_flash)
        
        btn_new.connect("clicked", self.on_new_clicked, self)
        btn_open.connect("clicked", self.on_open_clicked, self)
        btn_save.connect("clicked", self.on_save_clicked, self)
        btn_flash.connect("clicked", self.on_flash_clicked, self)

        self.box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL)

        #toolbutton = Gtk.ToolButton(stock_id=Gtk.STOCK_NEW)
        #toolbutton.set_is_important(True)
        #toolbutton.connect("clicked", self.on_new_clicked, self)

        #toolbar = Gtk.Toolbar()
        #toolbar.get_style_context().add_class(Gtk.STYLE_CLASS_PRIMARY_TOOLBAR)
        #toolbar.insert(toolbutton, 0)

        #self.box.add(toolbar)

        self.notebook = Gtk.Notebook()
        self.notebook.set_show_border(False)
        self.notebook.set_scrollable (True)
        Gtk.Notebook.popup_enable (self.notebook)
        self.notebook.connect("switch-page", self.on_page_changed, None)

        self.files = []
        self.create_sourceview()
        self.box.pack_start(self.notebook, True, True, 0)
        
        self.add(self.box)
        self.set_default_size(800,600)
        self.set_titlebar(self.header)
        self.set_icon_name("applications-development")
        self.show_all()

        mbtext = find_microbit()
        print(mbtext)

    def on_page_changed(self, happy, page, page_num, data):
        self.header.set_title(self.notebook.get_tab_label(page).get_text())
        logger.debug("Switched to tab with contents: " + page.bit_buffer.get_text(page.bit_buffer.get_start_iter(), page.bit_buffer.get_end_iter(), True))

    def create_sourceview(self, file = "/home/pi/projects/Bit/template.py"):
        logger.debug("Make Tab")
        filename = file.split('/')
        filename = filename[len(filename)-1]

        lm = GtkSource.LanguageManager.new()
        language = lm.guess_language(file, None)

        buffer = GtkSource.Buffer()

        if language:
            buffer.set_highlight_syntax(True)
            buffer.set_language(language)
        else:
            logger.warning('No language found for file "%s"' % file)
            buffer.set_highlight_syntax(False)

        source_file = GtkSource.File()
        source_file.set_location(Gio.File.new_for_path(file))
        source_file_loader = GtkSource.FileLoader.new(buffer, source_file)
        source_file_loader.load_async(GLib.PRIORITY_DEFAULT, None, None, None, None, None)

        sourceview = GtkSource.View.new_with_buffer(buffer)
        sourceview.set_auto_indent(True)
        sourceview.set_indent_on_tab(True)
        sourceview.set_show_line_numbers(True)
        sourceview.set_highlight_current_line(True)
        sourceview.set_smart_home_end(True)

        scrolledwindow = Gtk.ScrolledWindow()
        scrolledwindow.add(sourceview)

        tab_label = TabLabel(filename)
        tab_label.connect("close-clicked", self.on_close_clicked, self.notebook, scrolledwindow)
        
        self.notebook.append_page(scrolledwindow, tab_label)
        self.notebook.set_tab_reorderable(scrolledwindow, True)
        scrolledwindow.bit_file = source_file
        scrolledwindow.bit_buffer = buffer
        scrolledwindow.show_all()
        self.notebook.set_current_page(self.notebook.page_num(scrolledwindow))

    def on_close_clicked(self, widget2, sender, widget):
        logger.debug("Close Tab Requested")
        pagenum = self.notebook.page_num(widget)
        # Should check if saved
        self.notebook.remove_page(pagenum)

    def on_new_clicked(self, widget, data):
        logger.debug("New Tab Requested")
        self.create_sourceview()
        self.show_all()

    def on_save_clicked(self, widget, data):
        logger.debug("Save Requested")
        source_file_saver = GtkSource.FileSaver.new(self.notebook.get_nth_page(self.notebook.get_current_page()).bit_buffer, self.notebook.get_nth_page(self.notebook.get_current_page()).bit_file)
        self.notebook.get_tab_label(self.notebook.get_nth_page(self.notebook.get_current_page())).start_working()
        self.notebook.get_nth_page(self.notebook.get_current_page()).bit_buffer.set_modified(False)
        source_file_saver.save_async(GLib.PRIORITY_DEFAULT, None, None, None, self.done_io, self.notebook.get_nth_page(self.notebook.get_current_page()))
        
    def done_io(self, task, result, data):
        logger.debug("File IO Complete")
        self.notebook.get_tab_label(data).stop_working()
        
    def on_open_clicked(self, widget, data):
        logger.debug("Open Requested")
        filechooserdialog = Gtk.FileChooserDialog("Open", self, Gtk.FileChooserAction.OPEN, (Gtk.STOCK_CANCEL, Gtk.ResponseType.CANCEL, Gtk.STOCK_OPEN, Gtk.ResponseType.OK))

        filter_py = Gtk.FileFilter()
        filter_py.set_name("Python files")
        filter_py.add_mime_type("text/x-python")
        filechooserdialog.add_filter(filter_py)

        filter_any = Gtk.FileFilter()
        filter_any.set_name("Any files")
        filter_any.add_pattern("*")
        filechooserdialog.add_filter(filter_any)
        
        response = filechooserdialog.run()
        if response == Gtk.ResponseType.OK:
            file = filechooserdialog.get_filename()
            self.create_sourceview(file)
        filechooserdialog.destroy()
        
    def on_flash_clicked(self, widget, data):
        logger.debug("Flash Requested")

    def on_close_widow(self):
        logger.debug("Window closing")
        print(self.notebook.get_n_pages())
        for x in range(0, self.notebook.get_n_pages()):
            if (self.notebook.get_nth_page(x).bit_buffer.get_modified()):
                dialog = Gtk.MessageDialog(self, 0, Gtk.MessageType.QUESTION, Gtk.ButtonsType.YES_NO, "Unsaved Changes")
                dialog.format_secondary_text(self.notebook.get_tab_label(self.notebook.get_nth_page(x)).text + " has unsaved changs. Save them now?")
                response = dialog.run()
                if response == Gtk.ResponseType.YES:
                    print("QUESTION dialog closed by clicking YES button")
                elif response == Gtk.ResponseType.NO:
                    print("QUESTION dialog closed by clicking NO button")
                dialog.destroy()

class BitApp(Gtk.Application):

    def __init__(self, *args, **kwargs):
        super().__init__(*args, application_id="org.zanderbrown.bit",
                         flags=Gio.ApplicationFlags.HANDLES_COMMAND_LINE,
                         **kwargs)
        self.window = None

        self.add_main_option("test", ord("t"), GLib.OptionFlags.NONE,
                             GLib.OptionArg.NONE, "Command line test", None)

    def do_startup(self):
        Gtk.Application.do_startup(self)

        action = Gio.SimpleAction.new("about", None)
        action.connect("activate", self.on_about)
        self.add_action(action)

        action = Gio.SimpleAction.new("quit", None)
        action.connect("activate", self.on_quit)
        self.add_action(action)

        builder = Gtk.Builder.new_from_string(MENU_XML, -1)
        self.set_app_menu(builder.get_object("app-menu"))

    def do_activate(self):
        # We only allow a single window and raise any existing ones
        if not self.window:
            # Windows are associated with the application
            # when the last one is closed the application shuts down
            self.window = BitWin(application=self, title="Bit")
            self.window.connect("delete-event", self.on_quit)

        self.window.present()

    def do_command_line(self, command_line):
        options = command_line.get_options_dict()

        if options.contains("test"):
            # This is printed on the main instance
            print("Test argument recieved")

        self.activate()
        return 0

    def on_about(self, action, param):
        aboutdialog = Gtk.AboutDialog(transient_for=self.window, modal=True)
        aboutdialog.set_name("Bit")
        aboutdialog.set_version("1.0")
        aboutdialog.set_comments("Python Editor for Micro::Bit")
        aboutdialog.set_authors(["Alexander Brown"])
        aboutdialog.set_copyright("Copyright © 2016 Alexander Brown")
        aboutdialog.set_logo_icon_name("applications-development")
        aboutdialog.run()
        aboutdialog.destroy()

    def on_quit(self, action, param):
        self.window.on_close_widow()
        self.quit()



def run():
    setup_logging()
    app = BitApp()
    app.run(sys.argv)

if __name__ == "__main__":
    run()
